src/hotspot/share/runtime/arguments.cpp

```diff
@@ -2492,6 +2492,11 @@ jint Arguments::parse_each_vm_init_arg(const JavaVMInitArgs* args, bool* patch_m
         needs_module_property_warning = true;
         continue;
       }
+      if (match_option(option, "-Djava.system.class.loader=", &value) ||
+          match_option(option, "-Djdk.module.showModuleResolution", &value) ||
+          match_option(option, "-Djdk.module.validation", &value)) {
+        CDSConfig::handle_incompatible_property((const char*)option->optionString + 2);
+      }
```

Comment:
Since add_property can be called in multiple places, I think this check should be put inside add_property().

Also, for better modularity, the set of properties should be listed inside CDSConfig.cpp. Maybe something like this:

```
bool Arguments::add_property(...) {
  ...
  CDSConfig::check_incompatible_property(key, value); // checks the 3 properties inside
}
```

Reply:
Per our discussion offline, only call `CDSConfig::check_incompatible_property()` for `ExternalProperty`.

```
  if (internal == ExternalProperty) {
    CDSConfig::check_incompatible_property(key, value);
  }
```

I've pushed another commit based on your comment.